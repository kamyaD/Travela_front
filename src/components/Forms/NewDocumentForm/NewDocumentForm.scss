.add-file {
	height: 20px;
  width: 55px;
  color: #4F4F4F;
  font-family: "DIN Pro";
  font-size: 15px;
  line-height: 20px;
  margin-top: 20px;
  margin-bottom: 10px;
}

form fieldset:last-child .submit-area {
  width: 91%;
}

.document-input {
  box-sizing: border-box;
  height: 61px;
  width: 421px;
  border: 1px solid #D4D4D4;
  border-radius: 2px;
  background-color: #F8F8F8;
  margin-bottom: 40px;
  border-style: dashed;
  font-weight: 300;
  
  .file-errors {
    margin-top: 70px;
    color: #999999;
    font-size: 12px;
    display:flex;
    justify-content: space-between;
    
  }

  input[type=file] {
    height: 59px;
    width: 421px;
    opacity: 0;
    position: absolute;
    right: 28px;
    top: 123px;
    z-index: 3;
  }

  .document-upload-icon {
    height: 13.33px;
    width: 20px;
    position: absolute;
    top: 143px;
    left: 43px;
    z-index: 2;
  }

  .upload-file {
    height: 18px;
    width: 72px;
    color: #999999;
    font-family: "DIN Pro";
    font-size: 14px;
    line-height: 18px;
    position: absolute;
    left: 67px;
    top: 139px;
    z-index: 2;
  }
}

.modal {
  hr {
    position: relative;
    margin-top: 10px;
    margin-bottom: 2px;
    width: 100%;
    box-sizing: border-box;
    height: 1px;
    width: 479.6px;
  }
}

@import 'responsiveness'
