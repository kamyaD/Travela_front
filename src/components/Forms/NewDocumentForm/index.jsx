import React, { PureComponent } from 'react';
import {PropTypes} from 'prop-types';
import axios from 'axios';
import toast from 'toastr';
import { Input, FormContext, getDefaultBlanksValidatorFor } from '../FormsAPI';
import DocumentAPI from '../../../services/DocumentAPI';
import { errorMessage } from '../../../helper/toast';
import SubmitArea from './FormFieldSets/SubmitArea';
import documentUpload from '../../../images/icons/document-upload.svg';
import './NewDocumentForm.scss';

class NewDocumentForm extends PureComponent {
  constructor(props) {
    super(props);
    this.defaultState = {
      values: {
        name: '',
        file: ''
      },
      errors: {},
      hasBlankFields: true,
      isSubmitting: false,
    };
    this.state = {...this.defaultState};
  }

  validate = (field = 'name') => {
    let { values, errors } = this.state;
    [errors, values] = [{ ...errors }, { ...values }];
    let hasBlankFields = false;
    !values[field]
      ? (errors[field] = 'This field is required')
      : (errors[field] = '');

    hasBlankFields = Object.keys(values).some(key => !values[key]);
    const fileNameRegex = /(^[A-Za-z])(.+)$/;
    if(values[field] && !fileNameRegex.test(values[field])) {
      errors[field] = 'File name must start with an alphabet';
      // set to true to disable 'save' button
      hasBlankFields = true;
    }
    this.setState(prevState => {
      return { ...prevState, errors, hasBlankFields };
    });
    return !hasBlankFields;
  };

  handleCancel = () => {
    const { closeModal } = this.props;
    this.setState({ ...this.defaultState });
    closeModal();
  };

  handleUpload = (e) => {
    e.preventDefault();
    const file = e.target.files[0];
    const documents = ['image/jpeg', 'image/png', 'application/pdf'];
    const fileReader = new FileReader();
    if (!documents.includes(file.type)) {
      return toast.error('Incorrect file type uploaded');
    }
    if (file.size>Math.pow(10,7)) {
      return toast.error('Incorrect file size uploaded');
    }
    fileReader.onload = () => {
      this.setState(
        prevState =>
          ({ values:
            {
              ...prevState.values,
              name: (file.name).replace(/\.[^/.]+$/, ''),
              file: fileReader.result
            }
          }),
        this.validate
      );
    };
    fileReader.readAsDataURL(file);
  };

  handleSubmit = async event => {
    const { values } = this.state;
    this.setState({ hasBlankFields: true, isSubmitting: true });
    const { user, createDocument } = this.props;
    event.preventDefault();
    const formData = new FormData();
    formData.append('file', values.file);
    formData.append('upload_preset', process.env.REACT_APP_PRESET_NAME);
    try {
      let cloudinaryUrl;
      let publicId;
      delete axios.defaults.headers.common['Authorization'];
      if (values.file !== '') {
        const uploadedDoc = await axios.post(process.env.REACT_APP_CLOUNDINARY_API, formData);
        cloudinaryUrl = uploadedDoc.data.secure_url;
        publicId = uploadedDoc.data.public_id;
      }
      DocumentAPI.setToken();
      const documentData = {
        name: values.name,
        cloudinary_public_id: publicId,
        cloudinary_url: cloudinaryUrl,
        userId: user.result.userId
      };
      createDocument(documentData);
    } catch (error) {
      errorMessage('Unable to upload document');
      DocumentAPI.setToken();
      this.setState({ hasBlankFields: false });
    }
  }

  render() {
    const { errors, values, hasBlankFields, isSubmitting } = this.state;
    return (
      <FormContext
        targetForm={this}
        errors={errors}
        values={values}
        validatorName="validate" // uses default validator if this is undefined
      >
        <form onSubmit={this.handleSubmit}>
          <div className="add-file">Add File</div>
          <div className="document-input">
            <img src={documentUpload} alt="Document Upload" className="document-upload-icon" />
            <div className="upload-file">Upload File</div>
            <input type="file" name="file" onChange={this.handleUpload} id="upload-btn" />
            <div className="file-errors">
              <span>Allowed files: *jpeg, *pdf, *png</span>
              <span id="size">Max size: 10mb</span>
            </div>
          </div>
          <Input
            type="text"
            name="name"
            label="Name"
          />
          <SubmitArea hasBlankFields={hasBlankFields} onCancel={this.handleCancel} loading={isSubmitting} />
        </form>
      </FormContext>
    );
  }
}

NewDocumentForm.propTypes = {
  closeModal: PropTypes.func.isRequired,
};

export default NewDocumentForm;
