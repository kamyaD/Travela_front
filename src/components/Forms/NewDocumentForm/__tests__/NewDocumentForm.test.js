import React from 'react';
import { stub } from 'sinon';
import toast from 'toastr';
import NewDocumentForm from '..';

toast.error = jest.fn();

describe('<NewDocumentForm />', () => {
  let wrapper;
  const props = {
    user: {
      userId: 'getyHjk5',
    },
    createDocument: jest.fn,
    closeModal: jest.fn,
    onCancel: jest.fn,
  };

  const file = new Blob([{
    name: 'file.jpg',
    lastModified: 1517684494000,
    lastModifiedDate:
        'Sat Feb 03 2018 20:01:34 GMT+0100 (West Africa Standard Time)',
    webkitRelativePath: '',
    size: 212919,
    type: 'application/pdf'
  }], {type : 'application/pdf'});

  const invalidFileType = new Blob(['invalid file'], {type : 'text/plain'});

  beforeEach(() => {
    wrapper = mount(<NewDocumentForm {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the onClick prop when click is simulated', () => {
    const onCancel = jest.fn;
    wrapper.find('#cancel').simulate('click');
    expect(onCancel).toBeCalled;
  });

  it('should call the onChange prop when click is simulated', () => {
    const onChange = jest.fn;
    const event = {
      preventDefault: jest.fn(),
      target: {
        files: [
          file
        ]
      }
    };
    wrapper.find('#upload-btn').simulate('change', event);
    expect(onChange).toBeCalled;
  });

  it('should call onsubmit', () => {
    const onSubmit = stub()
      .withArgs('file', 'name');
    const uploadComponent = mount(<NewDocumentForm handleSubmit={onSubmit} /> );
    uploadComponent.simulate('submit');
    expect(onSubmit).toBeCalled;
  });

  it('shows a toaster if file type is invalid', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        files: [
          invalidFileType
        ]
      }
    };
    wrapper.find('#upload-btn').simulate('change', event);
    expect(toast.error).toHaveBeenCalledWith('Incorrect file type uploaded');
  });

  describe('NewDocumentForm.validate', () => {
    it('validates filename', () => {
      const instance = wrapper.instance();
      jest.spyOn(instance, 'validate');
      expect(instance.validate).toBeCalled;
    });

    it('validates required field', () => {
      wrapper.instance().validate();
      expect(wrapper.state('errors')).toEqual({ name: 'This field is required' });
    });

    it('validates filename', () => {
      wrapper.state('values').name = '12filename'; // set name field value to string starting with digits
      wrapper.instance().validate();
      expect(wrapper.state('errors')).toEqual({
        name: 'File name must start with an alphabet'
      });
    });
  });

});
