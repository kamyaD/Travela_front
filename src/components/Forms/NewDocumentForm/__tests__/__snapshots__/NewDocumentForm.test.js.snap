// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewDocumentForm /> renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="add-file"
  >
    Add File
  </div>
  <div
    className="document-input"
  >
    <img
      alt="Document Upload"
      className="document-upload-icon"
      src="document-upload.svg"
    />
    <div
      className="upload-file"
    >
      Upload File
    </div>
    <input
      id="upload-btn"
      name="file"
      onChange={[Function]}
      type="file"
    />
    <div
      className="file-errors"
    >
      <span>
        Allowed files: *jpeg, *pdf, *png
      </span>
      <span
        id="size"
      >
        Max size: 10mb
      </span>
    </div>
  </div>
  <div
    className="form-input  "
  >
    <label
      htmlFor="name"
    >
      Name
      <span
        style={
          Object {
            "color": "red",
          }
        }
      >
        *
      </span>
    </label>
    <input
      className=" "
      label="Name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <span
      className="error"
    />
  </div>
  <fieldset>
    <div
      className="submit-area"
    >
      <button
        className="bg-btn bg-btn--inactive"
        id="cancel"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="bg-btn bg-btn--active"
        disabled={true}
        id="submit"
        type="submit"
      >
        Add
      </button>
    </div>
  </fieldset>
</form>
`;
